File: native/src/activation.c
#include <include/activation.h>

float sigmoid(float x)
{
  return 1.0 / (1.0 + exp(-x));
}

float sigmoid_prime(float x, float error)
{
  return x * (1.0 - x) * error;
}

const Activation Sigmoid = {sigmoid, sigmoid_prime};

float tanh_prime(float x, float error)
{
  return (1.0 - (x * x)) * error;
}

const Activation Tanh = {tanf, tanh_prime};

float relu(float x)
{
  return x > 0.0 ? x : 0.0;
}

float relu_prime(float x, float error)
{
  return x > 0.0 ? error : 0.0;
}

const Activation Relu = {relu, relu_prime};

float linear(float x)
{
  return x;
}

float linear_prime(float x, float error)
{
  return error;
}

const Activation Linear = {linear, linear_prime};

float leaky_relu(float x)
{
  return x > 0.0 ? x : 0.01 * x;
}

float leaky_relu_prime(float x, float error)
{
  return x > 0.0 ? error : 0.01 * error;
}

const Activation LeakyRelu = {leaky_relu, leaky_relu_prime};

float relu6(float x)
{
  return x > 0.0 ? (x < 6.0 ? x : 6.0) : 0.0;
}

float relu6_prime(float x, float error)
{
  return x > 0.0 ? (x < 6.0 ? error : 0.0) : 0.0;
}

const Activation Relu6 = {relu6, relu6_prime};

float elu(float x)
{
  return x > 0.0 ? x : 0.01 * (exp(x) - 1.0);
}

float elu_prime(float x, float error)
{
  return x > 0.0 ? error : 0.01 * error * exp(x);
}

const Activation Elu = {elu, elu_prime};

float selu(float x)
{
  return x > 0.0 ? 1.0507 * x : 1.0507 * 1.67326 * (exp(x) - 1.0);
}

float selu_prime(float x, float error)
{
  return x > 0.0 ? 1.0507 * error : 1.0507 * 1.67326 * error * exp(x);
}

const Activation Selu = {selu, selu_prime};

const Activation *get_activation(ActivationType type)
{
  switch (type)
  {
  case ACT_SIGMOID:
    return &Sigmoid;
  case ACT_TANH:
    return &Tanh;
  case ACT_RELU:
    return &Relu;
  case ACT_LINEAR:
    return &Linear;
  case ACT_LEAKY_RELU:
    return &LeakyRelu;
  case ACT_RELU6:
    return &Relu6;
  case ACT_ELU:
    return &Elu;
  case ACT_SELU:
    return &Selu;
  default:
    return NULL;
  }
}

File: crates/core/src/cpu/activation.rs
use crate::Activation;
pub struct CPUActivation {
    pub activation: Activation,
    pub activate: ActivationFn,
    pub prime: ActivationFn,
}

type ActivationFn = fn(x: &f32) -> f32;

const ROOT_2_BY_PI: f32 = 0.7978845608028654;
const GELU_APPROX: f32 = 0.044715;

impl CPUActivation {
    pub fn from(activation: Activation) -> Self {
        let (activate, prime): (ActivationFn, ActivationFn) = match activation {
            Activation::Elu => (elu, elu_prime),
            Activation::LeakyRelu => (leaky_relu, leaky_relu_prime),
            Activation::Linear => (linear, linear_prime),
            Activation::Relu => (relu, relu_prime),
            Activation::Relu6 => (relu6, relu6_prime),
            Activation::Gelu => (gelu, gelu_prime),
            Activation::Selu => (selu, selu_prime),
            Activation::Sigmoid => (sigmoid, sigmoid_prime),
            Activation::Tanh => (tanh, tanh_prime),
        };

        Self {
            activation,
            activate,
            prime,
        }
    }

    pub fn from_option(activation: Option<Activation>) -> Option<CPUActivation> {
        if let Some(activation) = activation {
            Some(CPUActivation::from(activation))
        } else {
            None
        }
    }

    pub fn memoize_output(activation: &CPUActivation) -> bool {
        match activation.activation {
            Activation::Sigmoid | Activation::Tanh => true,
            _ => true,
        }
    }
}

fn sigmoid(x: &f32) -> f32 {
    return 1.0 / (1.0 + (-x).exp());
}

fn sigmoid_prime(x: &f32) -> f32 {
    return x * (1.0 - x);
}

fn tanh(x: &f32) -> f32 {
    return x.tanh();
}

fn tanh_prime(x: &f32) -> f32 {
    return 1.0 - tanh(x).powi(2);
}

fn linear(x: &f32) -> f32 {
    return *x;
}

fn linear_prime(_x: &f32) -> f32 {
    return 1.0;
}

fn relu(x: &f32) -> f32 {
    return x.max(0.0);
}

fn relu_prime(x: &f32) -> f32 {
    return if *x > 0.0 { 1.0 } else { 0.0 };
}

fn gelu(x: &f32) -> f32 {
    return (0.5 * x) * (1.0 + (ROOT_2_BY_PI * (x + GELU_APPROX * x.powi(3))).tanh());
}

fn gelu_prime(x: &f32) -> f32 {
    let tanned = (ROOT_2_BY_PI * (x + GELU_APPROX * x.powi(3))).tanh();
    return (0.5 * (1.0 + tanned))
        + (0.5 * x * (1.0 - tanned.powi(2))) * ROOT_2_BY_PI * (1.0 + 3.0 * GELU_APPROX * x.powi(2));
}

fn relu6(x: &f32) -> f32 {
    return x.max(0.0).min(6.0);
}

fn relu6_prime(x: &f32) -> f32 {
    return if *x > 0.0 && *x < 6.0 { 1.0 } else { 0.0 };
}

fn leaky_relu(x: &f32) -> f32 {
    return if *x > 0.0 { *x } else { x.max(0.01 * x) };
}

fn leaky_relu_prime(x: &f32) -> f32 {
    return if *x > 0.0 { 1.0 } else { 0.01 };
}

fn elu(x: &f32) -> f32 {
    return if *x >= 0.0 { *x } else { x.exp() - 1.0 };
}

fn elu_prime(x: &f32) -> f32 {
    return if *x > 0.0 { 1.0 } else { x.exp() };
}

fn selu(x: &f32) -> f32 {
    return if *x >= 0.0 {
        *x
    } else {
        1.0507 * (x.exp() - 1.0)
    };
}

fn selu_prime(x: &f32) -> f32 {
    return if *x > 0.0 { 1.0 } else { 1.0507 * x.exp() };
}
